### DISCUSSION TITLE
About Lidar accuracy

### DISCUSSION THREAD

**Discussion by ShuffleWire** [cyberbotics/webots#5362] [Q&A]
Hello !

I'm using a Lidar, and I'm facing the following issue : especially at high range (40meters), I seeing this behavior

![image](https://user-images.githubusercontent.com/83014015/196935233-bee0d974-b85c-4a2f-a1ba-c80ee8d81cdd.png)

This is a wall of a building, and the lidar is placed inside of the building.

We can see that many point are nor actually on the wall.

This behavior could be expected due to the usage of the depth buffer used to render the point cloud (I would say intuitively that there is some integer conversion going on). Am I right ?

If so, is there anyway to improve the Lidar accuracy ?

Thanks !

**Comment by omichel** [comment]
Yes, reducing the `Lidar.near` value should help to improve the accuracy.

**Reply by ShuffleWire** [reply]
It doesn't seems to make much of a difference... Taking example at two different ranges (about 5m and 40m) and near = 0.1 or 0.001
World have been reloaded after saving with the new near value each time.

At some closer range (5m)

near = 0.001
![image](https://user-images.githubusercontent.com/83014015/196939326-b84dbb49-06a4-47d5-ae66-8dfe0fec8a29.png)

near = 0.1
![image](https://user-images.githubusercontent.com/83014015/196939222-0c9596ab-c655-42c6-8fc7-5250240a24e8.png)

At some longer range (40m)

near = 0.001
![image](https://user-images.githubusercontent.com/83014015/196938620-09ab2018-6d8c-48ba-a2d7-7cdb39daa684.png)

near = 0.1
![image](https://user-images.githubusercontent.com/83014015/196938790-fd402d82-ea81-4016-83f0-853161e02c0f.png)

**Reply by omichel** [reply]
Well, then, I don't how to improve that...

**Comment by ShuffleWire** [comment]
Got a very strange behavior :  given this world :  
```
#VRML_SIM R2022b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"

WorldInfo {
}
Viewpoint {
  orientation -0.26608279967424403 -0.8814365690293144 0.39021727088296165 4.915925000582167
  position 18.277949915381868 -14.525300797004311 32.354988217620566
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Transform {
  children [
    Solid {
      translation 29.12 -22.83 0
      children [
        Shape {
          appearance PBRAppearance {
            roughness 1
          }
          geometry Box {
            size 10 10 10
          }
        }
      ]
    }
  ]
}
Robot {
  children [
    Lidar {
      horizontalResolution 1082
      fieldOfView 4.71225
      verticalFieldOfView 0.02
      numberOfLayers 1
      near 0.001
      minRange 0.5
      maxRange 40
    }
  ]
}
```
I could get this image :  
![image](https://user-images.githubusercontent.com/83014015/196944157-9dcf8b48-31bd-4702-8019-9fae479e1039.png)
Left side is not very well define, while the right side is very good !
Both side are roughly at the same distance from the lidar
Changing near from 1e-5 to 10 don't change anything.

**Reply by ShuffleWire** [reply]
Quick follows up : I've printed the value in Lidar::lidarImage() for this point cloud and got, excluding "inf"s :  

37.940495,37.720985,37.506813,37.001637,36.800098,36.602531,36.130146,35.944698,35.762695,35.320732,35.149670,34.982635,34.567368,34.410652,34.256447,33.866180,33.722084,33.349640,33.213970,33.080833,32.730984,32.605911,32.271122,32.153950,31.833387,31.723856,31.416870,31.314651,31.020773,30.924862,30.643461,30.554243,30.284412,30.200306,30.027296,30.124620,30.223589,30.323309,30.425129,30.527227,30.631222,30.736162,30.842348,30.950279,31.058941,31.170000,31.281431,31.394871,31.509272,31.625208,31.742987,31.861492,31.982721,32.104427,32.228287,32.353119,32.480640,32.608467,32.738766,32.870304,33.003407,33.138786,33.275169,33.414612,33.554512,33.697060,33.840900,33.986664,34.161140,34.313168,34.468349,34.629578,34.794876,34.965755,35.140446,35.321815,35.508442,35.698921,35.725552,35.925110,36.131096,36.341640,36.559544,36.600193,36.827198

Plotted, it give : (distance vs point position in cloud (ie ray angle))

![image](https://user-images.githubusercontent.com/83014015/199533227-c8336e26-dc92-4306-99f8-20ba415d8d67.png)

It show that it's an issue before that point in the code, I will continue to investigate

**Reply by ShuffleWire** [reply]
AFAIK, this data come from a memory mapped file, but I'm now stuck to now where does it come from...

**Reply by ShuffleWire** [reply]
I've checked that the resolution buffer was not at fault, (unlikely, considering that the right side of the data was very accurate).
`depthResolutionShader()` : `SHADER_DEPTH_RESOLUTION` is `WR_SHADER_PROGRAM_UNIFORM_TYPE_FLOAT`
`encodeDepthShader()` : `SHADER_ENCODE_DEPTH` is `WR_SHADER_PROGRAM_UNIFORM_TYPE_FLOAT`

so float should be okay, and in fact in the memory mapped file is effectively float data as well...
